sympy/solvers/ode/single.py

```diff
@@ -0,0 +1,754 @@
+#
+# This is the module for ODE solver classes for single ODEs.
+#
+
+from typing import ClassVar, Dict, Iterable, List, Optional, Type
+
+from sympy.core import Add, S, Mul, Pow
+from sympy.core.exprtools import factor_terms
+from sympy.core.expr import Expr
+from sympy.core.function import AppliedUndef, Derivative, Function, expand
+from sympy.core.relational import Equality, Eq
+from sympy.core.symbol import Symbol, Dummy, Wild
+from sympy.integrals import Integral
+from sympy.simplify import simplify, collect
+from sympy.simplify.radsimp import fraction
+from sympy.utilities import numbered_symbols
+from sympy.functions import exp
+
+from sympy.solvers.solvers import solve
+from sympy.solvers.deutils import ode_order, _preprocess
+from .ode import dsolve
+
+
+class ODEMatchError(NotImplementedError):
+    """Raised if a SingleODESolver is asked to solve an ODE it does not match"""
+    pass
+
+
+def cached_property(func):
+    '''Decorator to cache property method'''
+    attrname = '_' + func.__name__
+    def propfunc(self):
+        val = getattr(self, attrname, None)
+        if val is None:
+            val = func(self)
+            setattr(self, attrname, val)
+        return val
+    return property(propfunc)
+
+
+class SingleODEProblem:
+    """Represents an ordinary differential equation (ODE)
+
+    This class is used internally in the by dsolve and related
+    functions/classes so that properties of an ODE can be computed
+    efficiently.
+
+    Examples
+    ========
+
+    This class is used internally by dsolve. To instantiate an instance
+    directly first define an ODE problem:
+
+    >>> from sympy import Eq, Function, Symbol
+    >>> x = Symbol('x')
+    >>> f = Function('f')
+    >>> eq = f(x).diff(x, 2)
+
+    Now you can create a SingleODEProblem instance and query its properties:
+
+    >>> from sympy.solvers.ode.single import SingleODEProblem
+    >>> problem = SingleODEProblem(f(x).diff(x), f(x), x)
+    >>> problem.eq
+    Derivative(f(x), x)
+    >>> problem.func
+    f(x)
+    >>> problem.sym
+    x
+    """
+
+    # Instance attributes:
+    eq = None  # type: Expr
+    func = None  # type: AppliedUndef
+    sym = None  # type: Symbol
+    _order = None  # type: int
+    _eq_expanded = None  # type: Expr
+    _eq_preprocessed = None  # type: Expr
+
+    def __init__(self, eq, func, sym, prep=True):
```

Comment:
IMO, `__new__` should be used.

Reply:
This is not a Basic subclass so there is no need for `__new__`